tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/LocaleMapper.java

```diff
@@ -47,50 +41,30 @@
      *        {@link org.unicode.cldr.api.CldrDataType#SUPPLEMENTAL SUPPLEMENTAL} data.
      * @return IcuData containing locale data for the given locale ID.
      */
-    public static IcuData process(
-        String localeId,
-        CldrDataSupplier src,
-        Optional<CldrData> icuSpecialData,
-        PathValueTransformer transformer,
-        SupplementalData supplementalData) {
-
-        return process(
-            localeId,
-            src,
-            icuSpecialData,
-            transformer,
-            supplementalData.getDefaultCalendar(localeId));
-    }
-
-    @VisibleForTesting  // Avoids needing to pass a complete SupplementalData instance in tests.
-    public static IcuData process(
-        String localeId,
-        CldrDataSupplier src,
+    public static void process(
+        IcuData icuData,
+        CldrData unresolved,
+        CldrData resolved,
         Optional<CldrData> icuSpecialData,
         PathValueTransformer transformer,
         Optional<String> defaultCalendar) {
 
-        IcuData icuData =
-            new LocaleMapper(localeId, src, icuSpecialData, transformer)
-                .generateIcuData(localeId, true);
+        new LocaleMapper(unresolved, resolved, icuSpecialData, transformer).addIcuData(icuData);
         doDateTimeHack(icuData);
         defaultCalendar.ifPresent(c -> icuData.add(RB_CALENDAR, c));
-        return icuData;
     }
 
-    private final String localeId;
-    private final CldrDataSupplier src;
+    private final CldrData unresolved;
     private final Optional<CldrData> icuSpecialData;
 
     private LocaleMapper(
-        String localeId,
-        CldrDataSupplier src,
+        CldrData unresolved,
```

Comment:
Since we only want to generate data once, we pass in the data (which will cache and is reusable) rather than passing in the supplier (which generates new and independent each time).